File: src/mcp/inbox.ts
/**
 * Inbox MCP (§12): inbox.send (fire-and-forget note + push), inbox.ask
 * (multiple-choice question, pauses the session until answered), inbox.read.
 */
import { z } from "zod";
import { createMcpServer, type McpServer } from "./context.js";

export function createInboxMcpServer(): McpServer {
  return createMcpServer("inbox", [
    {
      name: "inbox.send",
      description:
        "Send a message to the human. Use ONLY when stuck or a decision is needed — never for routine progress.",
      inputSchema: z.object({ body: z.string().min(1) }),
      handler: async (rt, { body }) => {
        const msg = await rt.services.inbox.send({
          from: "agent",
          agentId: rt.manifest.agent.id,
          sessionId: rt.sessionId,
          taskId: rt.manifest.task?.id ?? null,
          goalId: rt.manifest.goal?.id ?? null,
          body,
        });
        await rt.onInboxNote(msg);
        return { ok: true, messageId: msg.id };
      },
    },
    {
      name: "inbox.ask",
      description:
        "Send a multiple-choice question to the human. The session pauses until the human answers; the answer resumes it.",
      inputSchema: z.object({
        body: z.string().min(1),
        choices: z.array(z.string().min(1)).min(2),
      }),
      handler: async (rt, { body, choices }) => {
        const msg = await rt.services.inbox.send({
          from: "agent",
          agentId: rt.manifest.agent.id,
          sessionId: rt.sessionId,
          taskId: rt.manifest.task?.id ?? null,
          goalId: rt.manifest.goal?.id ?? null,
          kind: "multiple-choice",
          body,
          choices,
        });
        await rt.onInboxQuestion(msg);
        return { ok: true, messageId: msg.id, paused: true };
      },
    },
    {
      name: "inbox.read",
      description: "Read the open messages in this session's thread.",
      inputSchema: z.object({}),
      handler: async (rt) => {
        const msgs = await rt.services.inbox.openForSession(rt.sessionId);
        return { messages: msgs };
      },
    },
  ]);
}

File: agentos/mcp/inbox.py
from __future__ import annotations
"""Inbox MCP. Port of src/mcp/inbox.ts."""
from agentos.mcp.context import McpServer, McpTool, create_mcp_server


def create_inbox_mcp_server() -> McpServer:
    async def inbox_send(rt, args):
        msg = await rt.services.inbox.send(
            from_="agent", agent_id=rt.manifest.agent.id, session_id=rt.session_id,
            task_id=rt.manifest.task.id if rt.manifest.task else None,
            goal_id=rt.manifest.goal.id if rt.manifest.goal else None,
            body=args["body"],
        )
        await rt.on_inbox_note(msg)
        return {"ok": True, "messageId": msg["id"]}

    async def inbox_ask(rt, args):
        msg = await rt.services.inbox.send(
            from_="agent", agent_id=rt.manifest.agent.id, session_id=rt.session_id,
            task_id=rt.manifest.task.id if rt.manifest.task else None,
            goal_id=rt.manifest.goal.id if rt.manifest.goal else None,
            kind="multiple-choice", body=args["body"], choices=args["choices"],
        )
        await rt.on_inbox_question(msg)
        return {"ok": True, "messageId": msg["id"], "paused": True}

    async def inbox_read(rt, args):
        msgs = await rt.services.inbox.open_for_session(rt.session_id)
        return {"messages": msgs}

    return create_mcp_server("inbox", [
        McpTool("inbox.send", "Send a message to the human (fire-and-forget).", inbox_send),
        McpTool("inbox.ask", "Send a multiple-choice question; session pauses until answered.", inbox_ask),
        McpTool("inbox.read", "Read open messages in this session's thread.", inbox_read),
    ])
